vortex-layout/src/segments/events.rs

```diff
@@ -0,0 +1,229 @@
+    Dropped(SegmentId),
+    Resolved(SegmentId),
+}
+
+impl Debug for SegmentEvent {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            SegmentEvent::Requested(req) => write!(
+                f,
+                "SegmentEvent::Registered({:?}, {})",
+                req.id, req.for_whom
+            ),
+            SegmentEvent::Polled(id) => write!(f, "SegmentEvent::Polled({:?})", id),
+            SegmentEvent::Dropped(id) => write!(f, "SegmentEvent::Dropped({:?})", id),
+            SegmentEvent::Resolved(id) => write!(f, "SegmentEvent::Resolved({:?})", id),
+        }
+    }
+}
+
+pub struct SegmentRequest {
+    /// The ID of the requested segment
+    id: SegmentId,
+    /// The name of the layout that requested the segment, used only for debugging.
+    for_whom: Arc<str>,
+    /// The one-shot channel to send the segment back to the caller
+    callback: oneshot::Sender<VortexResult<ByteBuffer>>,
+    /// The segment events that we post our resolved event back to.
+    events: Arc<SegmentEvents>,
+}
+
+impl SegmentRequest {
+    pub fn id(&self) -> SegmentId {
+        self.id
+    }
+
+    /// Resolve the segment request with the given buffer result.
+    pub fn resolve(self, buffer: VortexResult<ByteBuffer>) {
+        self.events.submit_event(SegmentEvent::Resolved(self.id));
+        if self.callback.send(buffer).is_err() {
+            // The callback may fail if the caller was dropped while the request was in-flight, as
+            // may be the case with pre-fetched segments.
+            log::debug!("Segment {} dropped while request in-flight", self.id);
+        }
+    }
+}
+
+impl SegmentEvents {
+    /// Get or create a segment future for the given segment ID.
+    fn segment_future(
+        self: Arc<Self>,
+        id: SegmentId,
+        for_whom: Arc<str>,
+    ) -> Shared<SegmentEventsFuture> {
+        loop {
+            // Loop in case the pending future has no strong references, in which case we clear it
+            // out of the map and create a new one on the next iteration.
+            match self.pending.entry(id) {
+                Entry::Occupied(e) => {
+                    if let Some(fut) = e.get().future() {
+                        break fut;
+                    } else {
+                        log::debug!("Re-requesting dropped segment from segment reader {}", id);
+                        e.remove();
+                    }
+                }
+                Entry::Vacant(e) => {
+                    let fut = SegmentEventsFuture::new(id, for_whom, self.clone()).shared();
+                    // Create a new pending segment with a weak reference to the future.
+                    e.insert(PendingSegment {
+                        id,
+                        fut: fut
+                            .downgrade()
+                            .vortex_expect("cannot fail, only just created"),
+                    });
+                    break fut;
```

Comment:
nit - break expressions are a bit esoteric IMO, will probably be easier to grok if it was just a return